### ISSUE TITLE
ReferenceArrayInput with AutocompleteArrayInput clearing input

### ISSUE THREAD

**Issue by KonkretneKosteczki** [marmelab/react-admin#8181]
I wanted to create AutocompleteArrayInput within ReferenceArrayInput,

**What you were expecting:**

I expected to be able to choose from the list elements just fine. 

**What happened instead:**

Whenever I start typing in the autocompleteArrayInput, when there is more than one item already selected, the choice list gets updated, then the value I have written into the input gets cleared, and the suggestion list gets updated again. It makes it almost impossible to use the form. 
![flickering](https://user-images.githubusercontent.com/46758933/191294297-eb5372a6-b78f-4736-a066-8a0f2829e2e9.gif)

**Steps to reproduce:**

I am using react-admin-firebase (not sure if that's important). I created a form with an input:
```
<ReferenceArrayInput
    source="qualifyingTasks"
    reference="Tasks"
>
    <AutocompleteArrayInput
        filterToQuery={(q) => ({name: q})}
        optionText={(record) => record.name}
    />
</ReferenceArrayInput>
```
It seems to be enough to reproduce the issue. Just keep adding tasks to the form and eventualy it will end up clearing the typed text. 

Can't seem to be able to reproduce the issue using the ra-data-fakerest, It might be related to react-admin-firebase library somehow.

**Environment**

* React-admin version: 4.3.3
* Last version that did not exhibit the issue (if applicable): -
* React version: 17.0.2
* Browser: Chrome
* Stack trace (in case of a JS error): -

**Comment by KonkretneKosteczki** [comment]
I seem to have managed to mitigate the issue by setting `clearOnBlur` option to false. Now the question is why does this happen exactly?

**Comment by PawelSuwinski** [comment]
I had the similar issue on react-admin 4.3.1. An update to last 4.3.4 version solved it in my case.

**Comment by PawelSuwinski** [comment]
After all I also ended up with `clearOnBlur={false}` to have robust solution. Also for that moment couldn't reproduce it using default CodeSanbox. I am using react-admin with api-platform.

**Comment by fzaninotto** [comment]
If you can't reproduce it with the codesandbox, it's possible that it's caused by a third-party data provider. Another possibility is that the CodeSandbox shows a small delay for dataProvider responses, and this is a race condition that only shows for larger delays.

In any case, we need a reproduction to start investigating this issue.

**Comment by magicxor** [comment]
@fzaninotto it seems that I have managed to reproduce this bug. Please check the following [CodeSandbox](https://codesandbox.io/p/github/magicxor/react-admin-repro-AutocompleteArrayInput-clearing-input-bug/draft/elegant-swartz?file=%2FREADME.md&workspace=%257B%2522activeFileId%2522%253A%2522clar358td0000l2hg21ptgzm2%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clar36ujl007s2a69175w52qb%2522%252C%2522isMinimized%2522%253Afalse%252C%2522path%2522%253A%2522%252F%2523%252Fuser%252F112%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522clar36r8h005r2a695rctdm7q%2522%252C%2522isMinimized%2522%253Atrue%257D%255D%257D) or the [repo](https://github.com/magicxor/react-admin-repro-AutocompleteArrayInput-clearing-input-bug).

Steps to reproduce the bug:
1) Open users -> user #112 (url path /#/user/112)
2) Start typing

**Comment by slax57** [comment]
I don't reproduce this issue in v4.6.1 in our storybook with the following (edited) story inside https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/input/AutocompleteArrayInput.stories.tsx :

```tsx
const BookEditWithReference = () => (
    <Edit
        mutationMode="pessimistic"
        mutationOptions={{
            onSuccess: data => {
                console.log(data);
            },
        }}
    >
        <SimpleForm>
            <ReferenceArrayInput reference="authors" source="author">
                <AutocompleteArrayInput
                    fullWidth
                    //optionText="name"
                    filterToQuery={q => ({ name: q })}
                    optionText={record => record.name}
                />
            </ReferenceArrayInput>
        </SimpleForm>
    </Edit>
);

export const InsideReferenceArrayInput = () => (
    <Admin dataProvider={dataProviderWithAuthors} history={history}>
        <Resource name="authors" />
        <Resource name="books" edit={BookEditWithReference} />
    </Admin>
);
```

Hence, I'll consider this issue fixed.

**Comment by magicxor** [comment]
Hi @slax57 Thank you for your reply, but I'm afraid that the issue is still here.

I have updated the repro of #8181 ([github](https://github.com/magicxor/react-admin-repro-AutocompleteArrayInput-clearing-input-bug), [codesandbox](https://codesandbox.io/p/github/magicxor/react-admin-repro-AutocompleteArrayInput-clearing-input-bug/draft/gallant-hamilton?file=%2Fpackage.json&workspace=%257B%2522activeFileId%2522%253A%2522clbq51ijz000w7odm7ejphs21%2522%252C%2522openFiles%2522%253A%255B%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clbq51n5z00vh2a6avpnb8gz8%2522%253A%257B%2522key%2522%253A%2522clbq51n5z00vh2a6avpnb8gz8%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clbq52v6w015s2a6ad0xjugvh%2522%252C%2522isMinimized%2522%253Afalse%252C%2522path%2522%253A%2522%252F%2523%252Fuser%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522clbq52t51013l2a6aocgig342%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clbq51n5z00vh2a6avpnb8gz8%2522%252C%2522spacesOrder%2522%253A%255B%2522clbq51n5z00vh2a6avpnb8gz8%2522%255D%257D)) and the repro of #8410 ([github](https://github.com/magicxor/react-admin-repro-AutocompleteArrayInput-clearing-input-bug), [codesandbox](https://codesandbox.io/p/github/magicxor/react-admin-repro-AutocompleteArrayInput-bug/draft/jolly-wildflower?file=%2Fsrc%2FUserAchievements.tsx&workspace=%257B%2522activeFileId%2522%253A%2522clbq50w6f00077nj66jxmc8ed%2522%252C%2522openFiles%2522%253A%255B%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clbq510so00e32a69ip3tyvyx%2522%253A%257B%2522key%2522%253A%2522clbq510so00e32a69ip3tyvyx%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clbq526zq00oe2a69m4fawvq7%2522%252C%2522isMinimized%2522%253Afalse%252C%2522path%2522%253A%2522%252F%2523%252Fuser%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522clbq5250q00m72a69wlaxooer%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clbq510so00e32a69ip3tyvyx%2522%252C%2522spacesOrder%2522%253A%255B%2522clbq510so00e32a69ip3tyvyx%2522%255D%257D))

**Comment by slax57** [comment]
@magicxor thanks for your reply and for the sandboxes.

Do you think you could provide a sandbox based on the [simple example](https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple)? I'm having a hard time working with sandboxes pointing to another project because it asks me for permissions to open them and I can't change the code easily.

Thanks

**Comment by slax57** [comment]
Besides, I had a quick look at the code from your reproduction project, and saw there are many places where you try advanced and custom features, that are beyond the use-cases we support officially.

For instance:
- Having an `<Edit>` inside a `<Show>`
- Overriding the `record` in the `<SimpleForm>` only
- Having the `achievements` stored actually on another resource than `user`
- Providing a custom `format` with no `parse`
- And so on...

It is still too complex and there are too many places prone to error for us to work with.

When taking your repo and just changing `UserEdit` with the following content it works flawlessly:

```tsx
export const UserEdit = () => {
  return (
    <Edit>
      <SimpleForm>
        <ReferenceArrayInput
          label="Achievements"
          source="achievements"
          reference="achievement"
          allowEmpty
          fullWidth
        >
          <AutocompleteArrayInput
            optionText={(o) => o.name}
            //format={formatFunc}
            filterToQuery={createSearchFilter}
            disableClearable
            clearOnBlur={false}
            fullWidth
          />
        </ReferenceArrayInput>
      </SimpleForm>
    </Edit>
  );
};
```

Hence, I'll still consider this issue fixed, unless you can provide a simpler repro, based on the simple example, and with supported (i.e. documented) use-cases only.

**Comment by magicxor** [comment]
@slax57 
> Do you think you could provide a sandbox based on the [simple example](https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple)? 

Sure. Please take a look: https://codesandbox.io/s/stupefied-cray-7n3gbo?file=/src/index.tsx

Basically, my API returns a post:
```json
        {
            id: 10,
            title: 'Totam vel quasi a odio et nihil',
            teaser:
                'Excepturi veritatis velit rerum nemo voluptatem illum tempora eos. Et impedit sed qui et iusto. A alias asperiores quia quo.',
            body:
                '<p>Voluptas iure consequatur repudiandae quibusdam iure. Quibusdam consequatur sit cupiditate aut eum iure. Provident ut aut est itaque ut eligendi sunt.</p><p>Odio ipsa dolore rem occaecati voluptatum neque. Quia est minima totam est dicta aliquid sed. Doloribus ea eligendi qui odit. Consectetur aut illum aspernatur exercitationem ut. Distinctio sapiente doloribus beatae natus mollitia. Nostrum cum magni autem expedita natus est nulla totam.</p><p>Et possimus quia aliquam est molestiae eum. Dicta nostrum ea rerum omnis. Ut hic amet sequi commodi voluptatem ut. Nulla magni totam placeat asperiores error.</p>',
            views: 721,
            average_note: 4.121,
            commentable: true,
            published_at: new Date('2012-10-19'),
            tags: [
                {
                    id: 2,
                    name: 'Technology',
                    published: false,
                },
                {
                    id: 4,
                    name: 'Photo',
                    published: false,
                }
            ],
            category: 'lifestyle'
        }
```

I want to edit the `tags` property of this post:
```tsx
export const formatFunc = (tags: any[]) => tags.map((tag: { id: number }) => tag.id);

export const parseFunc = (tagIds: number[]) => tagIds.map((id: number) => ({ id }));
```

```tsx
<ReferenceArrayInput
    reference="tags"
    source="tags"
    label="Tags"
>
    <AutocompleteArrayInput
        optionText={'name'}
        format={formatFunc}
        parse={parseFunc}
    />
</ReferenceArrayInput>
```

But search doesn't work properly in this input:
![firefox_rxDeIGk9b3](https://user-images.githubusercontent.com/8275793/208124222-5a13f773-7e3a-43a5-be60-f7a02dd4746c.gif)

**Comment by slax57** [comment]
@magicxor Thanks for the repro.

```js
            tags: [
                // my API returns objects, not just IDs:
                {
                    id: 2,
                    name: 'Technology',
                    published: false,
                },
                {
                    id: 4,
                    name: 'Photo',
                    published: false,
                }
            ],
```

> // my API returns objects, not just IDs:

That's actually the cause of the error. [`<ReferenceArrayInput>` only supports an array of ids as `source`](https://marmelab.com/react-admin/ReferenceArrayInput.html#source).

With `format` and `parse` you can only change the format of the value in the input element, but the value in the record needs to be of the expected shape.

If your API returns objects you probably need to tweak your dataProvider to transform them into an array of ids to be usable by `<ReferenceArrayInput>`.

**Comment by slax57** [comment]
Btw @magicxor , you might be interested in [this PR](https://github.com/marmelab/react-admin/pull/8511) if you are looking for a way to add lifecycle callbacks to your dataProvider :wink:

**Comment by woutb** [comment]
I still have this issue

> If your API returns objects you probably need to tweak your dataProvider to transform them into an array of ids to be usable by `<ReferenceArrayInput>`.

Adapting the dataProvider to only return an array of IDs solves this issue for me, but this deletes the name from my tags object, giving me a dropdown with undefineds.

```
[
    {
        "id": "59655158-669b-453b-8290-3a2280ab5856",
        "name": "Tag 1"
    },
    {
        "id": "69655158-e69b-453b-8290-3a2280ab5856",
        "name": "Tag 2"
    }
]
``` 

```typescript
getList(resource, params).then(( json: any ) => ({ data: json.data.map((tag: any) => {
                    return {
                        id: tag.code,
                    };
                }),
                total: json.total }));
``` 

```html
 <ReferenceArrayInput
          reference="tags" source="tags" name="tags"
          enableGetChoices={({q}) => q && q.length >= 2}
          sort={{field: "name", order: "ASC"}}
          clearOnEmpty={false}
  >
    <AutocompleteArrayInput optionText="name" name="tags"/>
  </ReferenceArrayInput>
```

**Comment by slax57** [comment]
@woutb It's becoming very hard to follow this issue. Can you please open a new issue if you believe you have found a bug?

Alternatively, if you need help troubleshooting your code, you can use one of these three options:

- Community support: Ask your question on [StackOverFlow](https://stackoverflow.com/questions/tagged/react-admin). This makes your question easy to find by the developer community. StackOverFlow has great SEO, gamification, voting, and reputation system.
- Community support: Ask your question on [Discord](https://discord.gg/B4G7AEQ3jm). Some react-admin users and contributors might have time to answer your questions.
- Professional support: If you're a [React-Admin Enterprise Edition](https://marmelab.com/ra-enterprise/#support) subscriber, you have unlimited access to the Professional Support service. This service guarantees that you'll get an answer shortly.
If your problem is urgent, please consider subscribing to the [Enterprise Edition](https://marmelab.com/ra-enterprise/), starting at 125€. It's a good way to contribute to the sustainability of react-admin.
